### PR TITLE
Add kvs-slow-log to KVS operations that return iterators

### PR DESCRIPTION
**Pull request by fsamuel-bs**
**Goals (and why)**: `kvs-slow-log` is an important log line that tells helps us to detect when the backing KVS is slow. This pr adds this logging to the methods that return iterators.

**Implementation Description (bullets)**: Track how much time it's taken to compute the `.next()` call takes on the iterator and log when it takes too long.

**Concerns (what feedback would you like?)**:
- Is this the correct place to track it?
- I think this works for CKVS. Does it also work for DbKVS?

**Where should we start reviewing?**: ProfilingKVSImpl

**Priority (whenever / two weeks / yesterday)**: By the next release.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2617)

### CODE REVIEW

**Comment by fsamuel-bs** [comment]
AFAICT it also works for DbKvs, as it seems to return an object built with `Iterators.concat(pageIterator)`

**Comment by fsamuel-bs** [comment]
Also, not sure if this is exactly what we want here: this will log if we have a `.next()` which takes more than 1 second. We might want to sum all the periods for an iterator and log if that's above 1 second.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 1]
```diff
@@ -390,7 +404,66 @@ public RowColumnRangeIterator getRowsColumnRange(TableReference tableRef,
...
+        RowColumnRangeIterator rowColumnRangeIterator = rowColumnRangeIteratorSupplier.get();
+        return new RowColumnRangeIterator() {
+            @Override
+            public Entry<Cell, Value> next() {
+                return maybeLog(rowColumnRangeIterator::next, logger);
+            }
+
+            @Override
+            public boolean hasNext() {
+                return rowColumnRangeIterator.hasNext();
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: do any of these `hasNext` impls actually load the next batch?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 2]
```diff
@@ -390,7 +404,66 @@ public RowColumnRangeIterator getRowsColumnRange(TableReference tableRef,
...
 
-    private  <T> T maybeLog(Supplier<T> action, BiConsumer<LoggingFunction, Stopwatch> logger) {
+    private RowColumnRangeIterator wrapRowColumnRangeIteratorWithLogger(
+            Supplier<RowColumnRangeIterator> rowColumnRangeIteratorSupplier,
+            BiConsumer<LoggingFunction, Stopwatch> logger) {
+        RowColumnRangeIterator rowColumnRangeIterator = rowColumnRangeIteratorSupplier.get();
+        return new RowColumnRangeIterator() {
+            @Override
+            public Entry<Cell, Value> next() {
+                return maybeLog(rowColumnRangeIterator::next, logger);
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: we probably want to update the log statement to indicate that this is one batch of the query - and maybe also indicate how many cells were in the batch

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 3]
```diff
@@ -160,7 +174,7 @@ public void delete(TableReference tableRef, Multimap<Cell, Long> keys) {
     @Override
     public void deleteRange(TableReference tableRef, RangeRequest range) {
         maybeLog(() -> delegate.deleteRange(tableRef, range),
-                logTimeAndTableRange("deleteRange", tableRef, range));
+                logTimeAndTableRangeOfIterator("deleteRange", tableRef, range));
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: why did this one change?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 4]
```diff
@@ -390,7 +404,66 @@ public RowColumnRangeIterator getRowsColumnRange(TableReference tableRef,
                         LoggingArgs.durationMillis(stopwatch)));
     }
 
-    private  <T> T maybeLog(Supplier<T> action, BiConsumer<LoggingFunction, Stopwatch> logger) {
+    private RowColumnRangeIterator wrapRowColumnRangeIteratorWithLogger(
+            Supplier<RowColumnRangeIterator> rowColumnRangeIteratorSupplier,
+            BiConsumer<LoggingFunction, Stopwatch> logger) {
+        RowColumnRangeIterator rowColumnRangeIterator = rowColumnRangeIteratorSupplier.get();
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: don't think you need to pass a supplier here. can probably pass the iterator itself

File: docs/source/release_notes/release-notes.rst [hunk 5]
```diff
@@ -48,8 +48,10 @@ develop
     *    - Type
          - Change
 
-    *    -
-         -
+    *    - |fixed|
+         - kvs.getRanges now also logs ``kvs-slow-log`` if the backing store is being slow to return the requested values.
```
Comment (nziebart) [inline docs/source/release_notes/release-notes.rst]: nit: probably `getRange` is the one people care about

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java (hunk 3 above)
Comment (fsamuel-bs) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: Refactored and didn't notice this had changed. Reverted.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java (hunk 4 above)
Comment (fsamuel-bs) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: Done.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java (hunk 2 above)
Comment (fsamuel-bs) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: The logger should have this information. Check above, where we construct it. I've made it more clear,

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java (hunk 1 above)
Comment (fsamuel-bs) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: They can. The backing store for at least some of these iterators is the AbstractIterator, which has:

```The initial invocation of hasNext() or next() calls this method, as does the first invocation of hasNext or next following each successful call to next. Once the implementation either invokes endOfData or throws an exception, computeNext is guaranteed to never be called again.```

I'll also wrap them.

File: docs/source/release_notes/release-notes.rst (hunk 5 above)
Comment (fsamuel-bs) [reply docs/source/release_notes/release-notes.rst]: :ack:

**Comment by nziebart** [comment]
After thinking about this more, I think this is actually a regression :(

In `KvsProfilingLogger#maybeLog`, we are creating a nontrivial `Monitor` object on every call to `next` and `hasNext`.

We may be better off adding slow-log warnings to the new thrift wrapper than trying to warn on these iterators, what do you think?

**Comment by nziebart** [comment]
After thinking about this more, I think this is actually a regression :(

In `KvsProfilingLogger#maybeLog`, we are creating a nontrivial `Monitor` object on every call to `next` and `hasNext`.

We may be better off adding slow-log warnings to the new thrift wrapper than trying to warn on these iterators, what do you think?

**Comment by fsamuel-bs** [comment]
Since we're instrumenting the Cassandra clients, I'm closing this in favor of #2651.

**Pull request closed without merging**